pyro/distributions/sine_skewed.py

```diff
@@ -0,0 +1,136 @@
+
+    as it would make the support for the Uniform distribution dynamic.
+
+    In the context of :class:`~pyro.infer.SVI`, this distribution can freely be used as a likelihood, but use as
+    latent variables it will lead to slow inference for 2 and higher dim toruses. This is because the base_dist
+    cannot be reparameterized.
+
+    .. note:: An event in the base distribution must be on a d-torus, so the event_shape must be (d,).
+
+    .. note:: For the skewness parameter, it must hold that the sum of the absolute value of its weights for an event
+        must be less than or equal to one. See eq. 2.1 in [1].
+
+    ** References: **
+      1. Sine-skewed toroidal distributions and their application in protein bioinformatics
+         Ameijeiras-Alonso, J., Ley, C. (2019)
+
+    :param torch.distributions.Distribution base_dist: base density on a d-dimensional torus. Supported base
+        distributions include: 1D :class:`~pyro.distributions.VonMises`,
+        :class:`~pyro.distributions.SineBivariateVonMises`, 1D :class:`~pyro.distributions.ProjectedNormal`, and
+        :class:`~pyro.distributions.Uniform` (-pi, pi).
+    :param torch.tensor skewness: skewness of the distribution.
+    """
+    arg_constraints = {'skewness': constraints.independent(constraints.interval(-1., 1.), 1)}
+
+    support = constraints.independent(constraints.real, 1)
+
+    def __init__(self, base_dist: TorchDistribution, skewness, validate_args=None):
+        assert base_dist.event_shape == skewness.shape[-1:], \
+            'Sine Skewing is only valid with a skewness parameter for each dimension of `base_dist.event_shape`.'
+
+        if (skewness.abs().sum(-1) > 1.).any():
+            warnings.warn("Total skewness weight shouldn't exceed one.", UserWarning)
+
+        batch_shape = broadcast_shapes(base_dist.batch_shape, skewness.shape[:-1])
+        event_shape = skewness.shape[-1:]
+        self.skewness = skewness.broadcast_to(batch_shape + event_shape)
+        self.base_dist = base_dist.expand(batch_shape)
+        super().__init__(batch_shape, event_shape, validate_args=validate_args)
+
+        if self._validate_args and base_dist.mean.device != skewness.device:
+            raise ValueError(f"base_density: {base_dist.__class__.__name__} and SineSkewed "
+                             f"must be on same device.")
+
+    def __repr__(self):
+        args_string = ', '.join(['{}: {}'.format(p, getattr(self, p)
+                                if getattr(self, p).numel() == 1
+                                else getattr(self, p).size()) for p in self.arg_constraints.keys()])
+        return self.__class__.__name__ + '(' + f'base_density: {str(self.base_dist)}, ' + args_string + ')'
+
+    def sample(self, sample_shape=torch.Size()):
+        bd = self.base_dist
+        ys = bd.sample(sample_shape)
+        u = Uniform(0., torch.ones(torch.Size([]), device=self.skewness.device)).sample(sample_shape + self.batch_shape)
```

Comment:
nit: `self.skewness.new_ones(())` would also get the correct dtype (float vs long)

Reply:
makes sense; I'll update it.